Blog post: 38% Left on the Table (HDA global optimum) (#191)
SECQUOIA/SECQUOIA.github.io: _posts/2026-08-16-hda-38-percent.md

---
layout: post
title: 38% Left on the Table
description: 'Revisiting the HDA process: a decade-long road from a GAMS internship to the certified global optimum of one of process synthesis'' founding models'
image: assets/images/hda-superstructure.png
date: 2026-08-16 08:00:00 -0400
---

<!-- markdownlint-disable MD033 -->

> *A decade with the hydrodealkylation process: from an internship at GAMS and a

Author: ```suggestion
*A decade with the hydrodealkylation process: from a visit to GAMS within a visit to CMU and a
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792708811 -->

feasibility pump for convex MINLP, through a disjunctive reformulation built
with Yunshan Liu, to the global optimum that the canonical model file never
mentioned.*

```text
--- the whole story, as a solver log ---
gamslib hda, DICOPT (1989-2026):      4322.55  integer optimal (OA)
2020, GDP port + logic-based OA:      5966.51  (in a slide deck)
2026, gdplib port, global solvers:    INFEASIBLE
2026, local solvers:                  "optimal" (violations up to 1.8e-2)
after the fix, 64-way enumeration:    5965.85  feasible, residual 1e-8
same flowsheet, pristine hda.gms:     5964.53
minlplib.org/hda, BARON dual bound:   5964.534  certified global
```

## An impromptu internship

> I started working on MINLP in 2015, in an impromptu internship at GAMS — the

Author: ```suggestion
I started working on MINLP in 2015, in an impromptu "internship" at GAMS within a visit to Prof. Ignacio Grossmann's research group at CMU, the
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792710538 -->

> beginning of a relationship with the company and its people that I remain

Author: ```suggestion
beginning of my PhD, a relationship with the company and its people that I remain
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792711271 -->

> grateful for, from that first project to the academic licenses that have

Author: ```suggestion
grateful for, from that first project to the support that have
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792711736 -->

> powered my group's research ever since. The internship turned into my first

Author: ```suggestion
powered my and my group's research ever since. The internship turned into my first
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792712086 -->

solver work: adding a feasibility pump to DICOPT for convex MINLP, eventually
published with Stefan Vigerske, Francisco Trespalacios, and Ignacio Grossmann
in [*Optimization Methods and
Software* (2020)](https://doi.org/10.1080/10556788.2019.1641498). My test
problems came from [MINLPLib](https://minlplib.org), and that is where I first

> kept running into `hda` — the hydrodealkylation of toluene, one of the

Author: ```suggestion
kept running into `hda`, the hydrodealkylation of toluene, one of the
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792712880 -->

founding examples of process synthesis. It appears in Douglas's 1988
*Conceptual Design of Chemical Processes*, became an MINLP superstructure in
Kocis and Grossmann's 1989 paper, and has shipped with every GAMS distribution
since, documented solution: profit 4322.55, reported by DICOPT as
integer-optimal.

I heard early on that something was off about `hda.gms`. But everything I was

> building lived in the convex world — outer approximation, feasibility pumps,

Author: ```suggestion
building lived in the convex world: outer approximation, feasibility pumps,
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792714630 -->

> methods whose optimality arguments lean on convexity — and `hda` is

Author: ```suggestion
methods whose optimality arguments lean on convexity. And `hda` is
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792715426 -->

aggressively nonconvex. My machinery had nothing honest to say about it. The
model went on a mental shelf labeled *someday*.

![HDA process superstructure](assets/images/hda-superstructure.png)

*The HDA superstructure (figure from Yunshan Liu's 2020 presentation): 72
streams, six discrete decisions — hydrogen feed purification, reactor type,
hydrogen recovery, methane recovery, and two liquid-separation choices —
embedded in one flowsheet with every alternative piped in parallel.*

## Where the nonconvexity lives

If you are a chemical engineer, you have written every equation in this
model. That is exactly why it is hard. MINLPLib classifies the instance as
indefinite, with 71 quadratic, 13 signomial, and 45 general nonlinear

> constraints — and each of those has a familiar face:

Author: ```suggestion
constraints, and each of those has a familiar face:
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792716713 -->

- **Arrhenius kinetics**, `k = 6.3e10 * exp(-Ea/(R*T))` — an exponential in
  temperature, multiplied into rate expressions.
- **Vapor pressure** via Antoine / Clausius-Clapeyron,
  `ln(pvap) = A - B/(T + C)` — a logarithm on one side, a pole in temperature
  on the other.
- **Flash equilibrium** — recoveries multiplying vapor pressures: bilinear
  and trilinear products.
- **Energy balances**, `Q = F * cp * dT` — flow times temperature, the
  classic bilinearity of every superstructure model.
- **Valves** — isentropic relations with fractional pressure powers,
  `T / p^((gamma-1)/gamma)`.
- **Reactor selectivity**, `1 - s = 0.0036 * (1 - X)^(-1.544)` — a signomial
  with a negative fractional power.
- **Shortcut distillation** (Fenske, Underwood) and the **Kremser equation**
  for absorption, `(1 - A^(N*eta)) / (1 - A)` — logarithms of flow ratios and
  an exponential in the tray count.

There is a second, sneakier consequence of the superstructure idea. These

> correlations describe *operating* equipment — but the optimizer must also

Author: ```suggestion
correlations describe *operating* equipment,  but the optimizer must also
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792717687 -->

evaluate them on flowsheets where the equipment is switched *off* and every
flow through it is zero. `ln(0)`, divisions by zero flow, `0^(-1.544)`: the
physics never goes there, but relaxations and reformulations do. This is why
formulations of such models sprout small epsilon constants inside logarithms

> and denominators — guards that keep the functions evaluable at the zero-flow

Author: ```suggestion
and denominators: guards that keep the functions evaluable at the zero-flow
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792718645 -->

points the algorithms visit. Hold that thought; one of those guards is the
villain of this story.

## A model everyone has met

HDA is not an obscure test case; it is closer to a rite of passage. Beyond
Douglas's book and the [Kocis and Grossmann
(1989)](https://doi.org/10.1016/0098-1354(89)85008-2) MINLP, it served among
the computational experience for DICOPT's development, became a flagship
example in [Türkay and Grossmann's
(1996)](https://doi.org/10.1016/0098-1354(95)00219-7) logic-based MINLP

> algorithms for process networks — the direct ancestor of the methods in this

Author: ```suggestion
algorithms for process networks, the direct ancestor of the methods in this
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792719366 -->

> story — and runs through the textbook literature descended from Douglas's

Author: ```suggestion
story, and runs through the textbook literature descended from Douglas's
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792719690 -->

treatment, such as Biegler, Grossmann, and Westerberg's *Systematic Methods
of Chemical Process Design*. In the modern era it lives on as `hda` in
[MINLPLib](https://minlplib.org/hda.html), appears in our own
[Pyomo.GDP paper](https://doi.org/10.1007/s11081-021-09601-7) ecosystem, and
anchors the [gdplib](https://github.com/SECQUOIA/gdplib) benchmark library.
Generations of algorithms have been graded against this flowsheet. That is
precisely what makes its documented solution worth getting right.

## 2020: the disjunctive port

> At Carnegie Mellon, Yunshan Liu — then an MS student I was advising —

Author: ```suggestion
At Carnegie Mellon, Yunshan Liu, then an MS student I was advising,
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792720577 -->

implemented HDA as a Generalized Disjunctive Program in Pyomo.GDP, turning
the six discrete decisions into proper disjunctions instead of big-M
constraints tangled through the algebra. The results we obtained together in
October 2020 were already remarkable, and in hindsight, prophetic:

> ![Big-M versus hull relaxation of a disjunction](assets/images/hda-bigm-vs-hull.png)

Author: This image is deceiving as it only shows the projection into the original space of variables.

Author: Besides the image corresponds to basic-steps and the Big-M / Hull discussion is kind of irrelevant here.

Author: Agreed on both counts; the figure and its caption are removed in f8674f5 (asset deleted as well).

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792721699 -->

*Why formulations matter: the same pair of disjunctions relaxed with big-M
(left) and with the hull reformulation (right) — the hull relaxation is
visibly tighter, at the price of more variables and constraints
(illustrations from the 2020 presentation).*

- ANTIGONE on the monolithic MINLP: **4048.39** — below even DICOPT's
  documented 4322.55.
- The same model rebuilt from the GDP via big-M: **4908.34**. Same problem,
  different formulation, different answer.
- Logic-based outer approximation on the GDP: **5966.51** — with a flowsheet
  nobody had recorded before: no hydrogen purification, adiabatic reactor,
  hydrogen recycle, methane recovered by membrane, and both separations done
  in columns.
- And the kicker: enumerating all 2^6 flowsheets and solving each NLP with

>   BARON — eight minutes of computation — *certified that solution as globally

Author: ```suggestion
  BARON after eight minutes of computation *certified that solution as globally
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792727971 -->

  optimal*.

![Logic-based outer approximation loop](assets/images/hda-loa-diagram.png)

*The logic-based outer approximation loop: an MILP master proposes a
flowsheet, a reduced NLP evaluates only the equipment that flowsheet actually

> uses — sidestepping the zero-flow singularities entirely — and OA cuts close

Author: ```suggestion
uses, sidestepping the zero-flow singularities entirely, and OA cuts close
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792725973 -->

## 2026: the model that could not be solved — or falsified

This year, a benchmark-health campaign across gdplib found the HDA port in a
strange state: global solvers (BARON, Gurobi) declared it *infeasible* in
seconds, while local solvers returned "optimal" solutions whose constraints,
re-evaluated directly, were violated by up to 1.8e-2. Every recorded HDA
benchmark result was an artifact. The model had drifted into genuine
infeasibility, and no one had noticed, because no one had asked a solution to
certify itself.

The diagnosis used a trick the port's fidelity made possible: it kept the
original's stream numbering. Solve the pristine `gamslib` model, transplant
its solution into the port variable by variable, rank our constraints by
violation. Everything held except one equation — an Antoine correlation whose
defensive epsilon inside the logarithm demanded a vapor pressure 1.3e-8
*below* its own Antoine-derived lower bound. The zero-flow guard, colliding
with a bound computed from the unguarded equation: unsatisfiable by one part
in 10^8. Large enough for global solvers to prove infeasibility; small enough
for local solvers to smear over.

> The repair replaced the guarded logarithms with exact, pole-free forms — a

Author: ```suggestion
The repair replaced the guarded logarithms with exact, pole-free forms, a
```
Also "pole-free" is not a term we use in this literature, we might need to describe/define it

Author: Suggestion applied, and "pole-free" is now defined in the same paragraph as shorthand: no denominator in these equations can reach zero anywhere the algorithms evaluate them, achieved by multiplying the offending equations through by their denominators. In f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792729270 -->

bounded log-vapor-pressure auxiliary defined by a bilinear equation, the
Arrhenius prefactor folded into its exponent, mole-fraction variables

> replacing regularized flow ratios — so the correlations stay evaluable at

Author: ```suggestion
replacing regularized flow ratios, so the correlations stay evaluable at
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792729640 -->

zero flow *without* perturbing the physics. The few epsilons that remain
guard true singularities (Fenske ratios as key-component flows vanish, the
Kremser expression at zero trays) and are documented as such. Along the way
we also learned that GAMS's model generator constant-folds variables pinned
by equal bounds with zero tolerance — it will reject `0 =E= 7.1e-15` as
infeasible — and collected three different "optimal, lower bound equals upper
bound" certificates from one global solver, at three different wrong values,

> each refuted by a feasible point we could verify to 1e-10.

Author: This last GAMS detail is not so important or necessary to put in the open

Author: Removed the GAMS constant-folding detail in f8674f5. The false-certificates observation stays, in a neutral and unattributed form.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792731064 -->

## Enumerate everything, again

With the model feasible, we re-ran our 2020 playbook with 2026 tooling: all
64 flowsheets, each fixed-configuration NLP solved with

> [POUNCE](https://github.com/jkitchin/pounce), John Kitchin's pure-Rust port

Author: ```suggestion
[POUNCE](https://github.com/jkitchin/pounce), pure-Rust port
```

Author: Applied verbatim in f8674f5.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792732225 -->

of Ipopt, driven from Pyomo over the AMPL NL interface — deliberately

> bypassing GAMS, so no translation layer could color the result. Minutes of

Author: This sounds like an attack against GAMS and it shouldn't be. Also we validates with other NLP solver (CONOPT and BARON back in the day)

Author: Reworded in f8674f5: the bypassing-GAMS framing is gone. The sweep is now presented as an independent solver stack complementing the CONOPT and BARON validations from 2020, with every code agreeing on the answer.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792733502 -->

infeasibilities. And here is the part that makes this a story about
scientific memory rather than a single bug: [MINLPLib's page for
`hda`](https://minlplib.org/hda.html) — the same library I mined for the
feasibility-pump paper a decade ago — reports a primal bound of -5964.534084
against a BARON dual bound of -5964.534089. In its minimization convention,
that is exactly the solution above, *globally certified*, sitting in public
view.

One scoping note, because the two models are siblings rather than twins. The
certificate belongs to the original MINLP, with the GAMS file's cost data.
The gdplib port follows the *paper's* cost data — which differs in a few
coefficients the port's own comments inventory (the absorber's fixed cost,
compressor costs, the electricity price) — so its 5965.85 is a verified
feasible solution and an exhaustive-over-flowsheets best, but not itself
globally certified. What transfers between the two is the thing a designer
cares about: *the same flowsheet wins in both*, and on that flowsheet — which
uses no absorber, sidestepping the largest cost difference — the two
objectives agree to 0.02%.

> To be clear about what this does and does not say about GAMS: nothing here is

Author: AI sounding

Author: Rewrote that paragraph in plainer first-person voice in f8674f5, and per the review body did a full pass removing every em-dash and the AI-flavored phrasing (e.g. "the kicker") throughout.

<!-- gh-arc:sha=f8674f576584c069a836f9869cd83604fe33f0b7:comment=3792735831 -->
